### PR TITLE
TDG Update and e2 fixes

### PR DESCRIPTION
**Pull request by NovaAstral**
- Fixed E2's being able to cloak the TARDIS when it was out of power or artron
- Made E2's stop being able to use functions on someone elses TARDIS if they have Isomorphic enabled

-- Time Distortion Generator Changes:
- Added Wire support to the TDG with Activate and Radius inputs and Active, Radius, and Health outputs
- Increased max size of TDG to 2048 (Default is still 1000)
- Added custom radius support for the TDG between 0 and 2048 using wire inputs
- Gave the TDG health that can deplete wether its active or not.
- Slightly increased efficiency of the Think in the TDG (It will only velocity the tdg if its unfrozen)
- Moved the TDG activate timer into an actual timer instead of think

### CODE REVIEW

File: lua/entities/gmod_tardis/modules/sh_cloak.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_cloak.lua]: This could be simplified to `return (self:GetData("power-state", false") and self:ToggleCloak()) and 1 or 0`

(I think, worth testing!)

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: Unless you have a good reason I think we can leave this to be `v:GetEnabled()` as we ideally shouldn't be directly accessing properties on other entities, same for the Radius should probably have a `GetRadius` function

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: Debugging print statement left in code

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: General comment about the if statements across all the changes they seem to have no space after the `if` and they have unnecessary brackets

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: I think to avoid this weird check (this confused me for a solid few minutes until I worked out why it's here) it would be better to replace the radius parameter of this function with an `outside` boolean parameter and then you can reference the `search_radius` if it's true or the `ent.metadata.Interior.ExitDistance` when it's false. You'll probably also want to pass in the entity itself as the first parameter in order to access the metadata

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: Formatting is a little off here

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: After the `return end` you don't need the `else` section you can just continue the code normally here

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: Is this `Wire_TriggerOutput` call needed? It looks like it's already called inside the TurnOn function

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: It would be good to have a `SetRadius` function similar to the TurnOn function to do this so it could be used by other things in future

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: Usage of timers is dangerous in GMod as they are very error prone and unreliable, can we return this to using the CurTime + Think function method as before?

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [inline lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: It would be good to rearrange this so that `self` was passed in as the first argument and then we could grab the `player` off that as well as pass it into `ent:HandleE2(name, self, ...)` like so, rather than passing in `self.player` and then `self` later on in the same call.

Comment (AmyJeanes) [inline lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: This is more of a comment for myself but we need to handle how this will interact with the legacy addon, currently there is effectively a copy of this file in the legacy addon but I'm thinking it might just be better to remove it and require the normal addon to use Wiremod against both legacy and the rewrite - otherwise it is a pain to keep updating legacy to keep it in sync with the rewrite for E2.

**Review by AmyJeanes — CHANGES_REQUESTED** [review]
Hey Nova, many thanks for contributing! I like the general theme of your changes here and more wiremod support is always welcome, we have definitely been lacking in that department historically when adding new features!

I hope I'm not scaring you off with I believe your first contribution to this addon with my comments, I don't mean them in a bad way at all and it's really great to see someone else making changes to the core addon!

If you want to discuss any of them outside GitHub you can feel free to message me on Discord, you already know my username as you asked me about how to create a branch earlier.

I know @francodml in particular will be pleased to see further Wiremod support!

Again nice work, and for a first contribution especially this is great stuff!

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [reply lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: On further thinking actually I think it might be best left to the HandleE2 function to deal with this which already has the player passed in as part of the E2 context.

I think the GetX functions are okay to run against any TARDIS regardless of isomorphic for example which we can only really handle in that function. There is an `ENT:CheckSecurity(ply)` function you can use to check permissions for this for things like `SetFlight` etc

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (NovaAstral) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: I don't have a good reason, I accidentally'd the GetEnabled and only noticed right before commit

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (NovaAstral) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: I personally find it easier to read this way, and it just feels better (probably because I started coding as an E2 user) but I can change it if you'd like

Comment (AmyJeanes) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: If you could please - thanks, then it matches Lua conventions and the conventions of the TARDIS codebase

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (NovaAstral) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: I'm not quite sure how to do this

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: Fixed

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: Fixed

File: lua/entities/gmod_tardis/modules/sh_time_distortions.lua
Comment (AmyJeanes) [reply lua/entities/gmod_tardis/modules/sh_time_distortions.lua]: As discussed on Discord I've done this bit for you

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: Formatting is a little off here, looks like you've got a tab character instead of spaces and there's a space after the left bracket

File: lua/entities/gmod_tardis/modules/sh_cloak.lua
Comment (AmyJeanes) [inline lua/entities/gmod_tardis/modules/sh_cloak.lua]: You should use `self:GetPower()` here instead of checking the power state directly

File: lua/entities/gmod_time_distortion_generator/init.lua
Comment (AmyJeanes) [inline lua/entities/gmod_time_distortion_generator/init.lua]: This should probably call it with self.Radius instead of hardcoding to 1000

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [inline lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: We probably don't want `self.player` on this one

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [inline lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: How does this behave in E2 if it's expected to get something back but returns nothing now - does it cause any issues?

File: i18n/languages/en.json
Comment (AmyJeanes) [inline i18n/languages/en.json]: Could we change the new strings to say Expression2 rather than E2 for clarity?

**Review by AmyJeanes — CHANGES_REQUESTED** [review]
This is looking great and almost perfect! Just a couple of small things to sort out before we can merge :)

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [reply lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: Looks good!

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (NovaAstral) [reply lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: I don't actually even know how that happened

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (NovaAstral) [reply lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: From what I tested, the E2 just does nothing

**Review by AmyJeanes — APPROVED** [review]
LGTM! Nice work 🙂

File: lua/entities/gmod_wire_expression2/core/custom/tardis2.lua
Comment (AmyJeanes) [reply lua/entities/gmod_wire_expression2/core/custom/tardis2.lua]: I'll handle this before deployment

**Pull request merged**
